include/nbl/builtin/hlsl/mpl.hlsl

```diff
@@ -24,21 +24,24 @@ struct clz_masks
 {
     static const uint16_t SHIFT = uint16_t(1)<<(bits_log2-1);
     static const uint64_t LO_MASK = (1ull<<SHIFT)-1;
-};
 
-template<>
-struct clz_masks<0>
-{
-    static const uint16_t SHIFT = 0;
-    static const uint64_t LO_MASK = 0;
+    // static const uint16_t SHIFT = type_traits::conditional<bool(bits_log2),type_traits::integral_constant<uint16_t,uint16_t(1)<<(bits_log2-1)>,type_traits::integral_constant<uint16_t,0> >::type::value;
+    // static const uint64_t LO_MASK = type_traits::conditional<bool(bits_log2),type_traits::integral_constant<uint64_t,(1ull<<SHIFT)-1>,type_traits::integral_constant<uint64_t,0> >::type::value;
 };
 
+// template<>
+// struct clz_masks<0>
+// {
+//     static const uint16_t SHIFT = 0;
+//     static const uint64_t LO_MASK = 0;
+// };
+
 template<uint64_t N, uint16_t bits_log2>
 struct clz
 {
     static const bool CHOOSE_HIGH = N&(clz_masks<bits_log2>::LO_MASK<<clz_masks<bits_log2>::SHIFT);
     static const uint64_t NEXT_N = (CHOOSE_HIGH ? (N>>clz_masks<bits_log2>::SHIFT):N)&clz_masks<bits_log2>::LO_MASK;
-    static const uint16_t value   = type_traits::conditional<bits_log2,clz<NEXT_N,bits_log2-1>,type_traits::integral_constant<uint16_t,0> >::type::value + (CHOOSE_HIGH ? 0ull:clz_masks<bits_log2>::SHIFT);
+    static const uint16_t value = type_traits::conditional<bits_log2,clz<NEXT_N,bits_log2-1>,type_traits::integral_constant<uint16_t,0> >::type::value + (CHOOSE_HIGH ? 0ull:clz_masks<bits_log2>::SHIFT);
```

Comment:
can you make that `conditional_value` so we can shorten expressions like this?

Also I'd still have a `clz_masks` and have the conditional live there.